### PR TITLE
ClientModel: Add BufferResponse to RequestOptions

### PR DESCRIPTION
**Pull request by annelo-msft**
Some service endpoints such as https://platform.openai.com/docs/api-reference/chat/create can return response content in different formats based on values sent in the request.  Given the way System.ClientModel-based clients model each endpoint with a single protocol method, we want to be able to let the protocol method caller specify whether or not the pipeline should buffer the content for a given response, based on what input they sent to the service.

This PR adds an optional BufferResponse property to RequestOptions.  If a protocol-method caller sets the value for BufferResponse, it will override any value set in the client code.

### CODE REVIEW

File: sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs [hunk 1]
```diff
@@ -239,6 +239,7 @@ public abstract partial class PipelineTransport : System.ClientModel.Primitives.
     public partial class RequestOptions
     {
         public RequestOptions() { }
+        public bool? BufferResponse { get { throw null; } set { } }
```
Comment (KrzysztofCwalina) [inline sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:242]: why is this nullable? Should we simply have true as the default?

Comment (annelo-msft) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:242]: If it always has a value, I wasn't sure how to compose it with the existing value on message.BufferResponse.  

Did you have a solution to that that I didn't think of?

Comment (KrzysztofCwalina) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:242]: What is the problem with the composition? I was thinking that the last set to message.BufferResponse wins

Comment (annelo-msft) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:242]: the shape of the request creation helper looks like this:

```csharp
      var message = _pipeline.CreateMessage();
      var request = message.Request;
      message.BufferResponse = false;
      message.Apply(options);  // <-- sets buffer response from options
```

If options.BufferResponse defaults to true, then a client cannot specify an operation that doesn't buffer the response.

Comment (KrzysztofCwalina) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:242]: Clients have protocol methods and convenience methods. RequestOptions are a parameter to protocol methods. Clients should not be picking whether protocol methods buffer or not. The default should be to buffer and RequestOption can change that. So I am not sure when clients "specify" that operation does not buffer.

Comment (annelo-msft) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:242]: Discussed offline: I believe this approach -- where BufferResponse is nullable -- will work without changes to the generator today.   @KrzysztofCwalina is concerned that a tri-state property complicates things for only a small benefit.

We agreed that, given short timeline needs for this feature, we can merge this as-is to unblock unbranded client development and hopefully return to revisit to understand what generator changes would be required to support the non-nullable bool property, with the story: "for protocol methods, we buffer by default. If the user does not like it, they can change the option. Convenience methods set the option however they want."

**Review by KrzysztofCwalina — APPROVED** [review]

**Pull request merged**